examples/ffi/profiles.c

```diff
@@ -14,15 +14,15 @@ int main(void) {
   const ddog_prof_Slice_ValueType sample_types = {&wall_time, 1};
   const ddog_prof_Period period = {wall_time, 60};
 
-  ddog_prof_Profile_NewResult new_result = ddog_prof_Profile_new(sample_types, &period, NULL);
-  if (new_result.tag != DDOG_PROF_PROFILE_NEW_RESULT_OK) {
+  ddog_prof_Result_HandleProfile new_result = ddog_prof_Profile_new(sample_types, &period, NULL);
+  if (new_result.tag != DDOG_PROF_RESULT_HANDLE_PROFILE_OK_HANDLE_PROFILE) {
```

Comment:
I don't write C API these days, but this looks worse. Can we use renames and/or aliases to improve this? Examples:

- `ddog_prof_Profile_NewResult` aliased to `ddog_prof_Result_HandleProfile`
- `ddog_prof_Profile` aliased to `ddog_prof_Handle_Profile`

Not sure what to do about the enum. Maybe @ivoanjo has thoughts?

Reply:
Hmmm yeah I think Levi has a good point here, the "handle" change does look a bit like libdatadog internal implementation details being leaked across the FFI.

E.g. looking at it from the pure API client side, I'm not sure it's relevant for client to care what a `ddog_prof_Profile` actually is: is it a profile? a handle? something else?

I also know that (at least in the past) some of our types were a bit less what we wanted because of cbindgen limitations, and I think it's still a reasonable trade-off to have the types uglier if the rust code becomes less error-prone/etc. But ideally having both (clear naming for ffi and better rust types) would be great.

> ddog_prof_Profile aliased to ddog_prof_Handle_Profile

TL;DR: :+1: on this one :)

---

On the `Result` change...

We've been using the convention of calling `Module_operation` returning `Module_OperationResult` since [this PR](https://github.com/DataDog/libdatadog/pull/73). (And [we did explore a few more options at the time](https://docs.google.com/spreadsheets/d/1iuuSCJ0NcSOpo2RVvkQ8f8ugz2XlEbbpFZLVOg6WioE/edit?gid=0#gid=0))

One advantage of this new convention is that it may result (no pun intended) in fewer `...Result` types, because they are now named by their payload's type, rather than the operation's type.

I still think I slightly favor more the current convention, but I think both seem quite reasonable. Here's a few examples for a comparison:

| **Current convention**                   | **with this PR's convention**                                    | **Notes** |
|----------------------------------        |----------------------------------                                |-----------|
| `ddog_prof_Profile_NewResult`            | `ddog_prof_Result_HandleProfile`                                 | |
| `ddog_prof_Exporter_NewResult`           | `ddog_prof_Result_Exporter`                                      | |
| `ddog_prof_Exporter_Request_BuildResult` | `ddog_prof_Result_Request`? `ddog_prof_Result_Exporter_Request`? | How to handle the deeper namespace?
| `ddog_prof_Exporter_SendResult`          | `ddog_prof_Result_HttpStatus` ? `ddog_Result_HttpStatus` ?       | How to handle when the type in the result is not named after the operation? |
| `ddog_prof_Profile_Result`               | `ddog_prof_ResultBool`? `ddog_ResultBool`?                       | (Similar to above) |
| `ddog_prof_Profile_SerializeResult`      | `ddog_prof_Result_EncodedProfile` ||

(There's also a few more results in ddcommon-ffi)

**Not wanting to discourage changes** (_I think one of the key ingredients of libdatadog's massive success so far has been having "no sacred cows", and instead always striving to leave things better, even if in the very short term it causes a bit of migration pain_) I'll suggest that if we plan to introduce the new convention, I don't think we should mix both -- I think we should rip-off the band-aid and convert everything to the new convention.